### PR TITLE
Lesson 3

### PR DESCRIPTION
**Pull request by Buzyuka**
https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&page=8&title=alg_3#R%3Cmxfile%3E%3Cdiagram%20id%3D%22a6bf7b3d-ddef-8e7a-10f4-2a949a4a8c7c%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%201%22%3E7VpLc%2BMoEP41qkoOM6VnYh1jJ9k5zFSlNqnanSO2iMQsFh6EYzu%2FfhsJPcFOPJHkZHZ9UEHTIGi%2B%2FrpBtrzZcvsHR6vkG4swtVw72lreteW6YWDDUwp2hSAIJ4Ug5iQqRE4tuCfPWAlVv3hNIpy1FAVjVJBVW7hgaYoXoiVDnLNNW%2B2R0fZbVyjGmuB%2Bgagu%2FYtEIimkE%2Feyln%2FBJE7KNzsXYdEyR4t%2FYs7WqXqf5XqP%2Ba9oXqJyLLXQLEER2zRE3o3lzThjoigttzNMpWlLsxX9bve0VvPmOBWv6uCqeYhduXYcgSlUlXGRsJiliN7U0mm%2BPixHsKGWiCWFogNFvCXibyn%2BHKja97IlFXzXaJLV72oAnEZXcs%2BguqAoy8iiEN4SWg78AwuxUzBBa8FAVE%2FtK2MrpVcsRq5grz2UKGNrvlBaygYC8RgrLa%2FaC4A4ZksM8wUVjikS5Kk9OlJgiyu92uBQUDY321%2B9%2BgnRtRrUApuE1%2FI5tS1Y1OSyLMNzmj9vLPeCwkSncw6lWFQLb%2B4ipeAscrc2CRH4foXy1W7AXdt7piaAucDbw0bTzVF2KNGsvN1X1U3tOpVK0nCbUu8tBrwYBL8N9DYAWzVV6FVtJ8avp%2BM3eCN%2B866wKrRrKKwYSUXWGPlOCmocTNowcPwO9xxW99tUBYXi%2FTUMqoW8ChmONxa1vb%2Ft90eiL0%2BjLyJbU%2FkulMb4DFphENuxz7XNgNC3ksUEbxEYApa%2FwpzAFDCvpXelyD0Jk1X1Majscgi8OvupzHmHVOb3H4p%2Ficq65OS%2BwGUvcN%2FbycweAhwfLE8zgCMcieh8jei2luSBQOrIEsik2Nb3iSdsOV9n4%2FCX34GtrfPXxEBfkx7oK9BMxLFYcxkLbhGFVPR9pagjEvtEs0wjyQ%2FyJF8%2FhsFaRdsMHGfkGc1zBemPisNAO5haAeBvKl0uK7xPdkCUxBBXryl%2BlENJAxI4414psZDuOM3A7CSNH3Lf%2FOT3Y263Y%2B6Kvxrm9g3Wdnuwdrgfhw98%2Fb5g6I2ZXziOGYe%2BOmD%2BVgh0wg4CJ6FmaW8gBDqDZHIHgvXhTC5lKW5FavuISB2hLMln9eth2zHEbXekuO3ogVvbm%2FIgAssQBNE%2F8ULA2YW%2BghHmTAi2LJ1DgUiWc2Dndm6pPIL9Z4wyXu8KbLsA8M%2FAgJjLvfxU7B9nAiwhj0XXod2PS3h2xyUuNY8og3grN%2BhhF%2FTcICJPJJJUcrbVj4Yfmnk8x%2F0cdKKfZmhTEtYH9Rjg%2Fd87J5SGbBHOnk3r%2F0bXNkfZF690T5iW%2BJ1jg68D1piVeGEP9gr%2BR2z9kauJ2LEu8VxDXqg%2BMABBpuUnhgz%2FVAddCYWiHYZuqpi6QcyTN3qv73BWhYYjOhXXjEd0yK8nj5nVEbrukBN37GO0z89fUD7wCantZMOzkNe9QzMcjoynxj4OR8Pcvn40GjJl6hdj0ZApVTc4jAymbh5Gi9DpqgAqn5DCXenIpjgdyNMpycQxQwOHvsUjGycVRCmmLOZoabU%2FnLTa7uqGE6QRptvHwW43yi%2FB3bzrqgGP5o2b4cPxCTOu6kAwgKmgWv%2BhpPj2UP9px7v5Fw%3D%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%225ea71882-e04f-befe-bfe6-bd2e726e53fc%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%202%22%3E7VdLj9owEP41kdpLBQ7hceS17aGVKnFo9%2BhNhsStE0eOWaC%2FvuN4QpINuyWtgEs5WONvxuN5fHaM5y%2FTw0fN8%2BSLikB6bBAdPH%2FlMTYLBjha4OiAYDZ1QKxF5KBhDWzELyCQ1sU7EUHRMjRKSSPyNhiqLIPQtDCutdq3zbZKtnfNeQwdYBNy2UW%2FicgkDp2ySY1%2FAhEn1c7D8cxpnnj4M9Zql9F%2BHvO35c%2BpU175okSLhEdq34D8tecvtVLGSelhCdKWtiqbW%2FfwivYUt4bMXLIgoDDMsUodIqwETZU2iYpVxuW6RhdlemAdDHCWmFSiOEQRDsJ8t%2FCHgGaPlSYz%2BthQ2ekjOYAsmtuW4TSUvChE6MAHISvHP8CYI7GE74xCqA7ts1I52blkbAavloOgQu10SFaMCMZ1DGTln1qBDAeVAsaLJhokN%2BK57Z0T1%2BKTXV1vFKjk58tPWz9zuSOnHtZktrLjYuBhUtNJJeO4KMd1t2dS4smwvdknwsAm52Vuezyb7Q7RdqANHN4uUTd5WuCPibp0tEc03dfnZFixO2mckcruX8o1%2Fs9WR802W0c3Yqt%2Fnq3luGDETTuOPH%2FusbHE%2BBZPGqXYSrIwXbDTUbwRcytiGbiUIFWseYqGOWiBEYN%2BqftaK%2B5A%2FxPXb8H%2F6a34f2eOj7ocn9yI46Mux4mzRc6zirUii0SIDxRstLUZeMFC4OPH9t0Z4zZN%2B3M%2Btkr3W1DucLG1x5b93PcMX2T97CHbpaC5gR5r3uGV8b5nWNu%2FyTrot4j1M7cccTS5fE0v%2F8HqD9Zv3Lft%2B%2BH6Fyh7cYGyQfcC9a91gU7Of8HmjW%2BXe3sF5dvrzHvjji8tNr3epwan9Z%2BOUtf4Y%2BevfwM%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%22ceb3e023-e3be-b663-f6ef-c709c21a45ea%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%203%22%3E7Vrdb5swEP9rkNqXKZjwkccmTdaHTZrUh61Pkxtc8Gowcpwm2V%2B%2FA2wwgXRJmo%2B1S1Uh%2B3w%2B2%2Bff74yPWM4oWX4WOIu%2F8pAwC%2FXCpeXcWggN3B48c8GqFLiDoBREgoalyK4F9%2FQ3UULVL5rTkMwaipJzJmnWFE55mpKpbMiwEHzRVHvirDlqhiPSEtxPMWtLv9NQxqU0QH4tvyM0ivXItjcoWx7x9DkSfJ6q8SzkPBV%2FZXOCtS210FmMQ74wRM7YckaCc1mWkuWIsNy12m1lv8mG1mregqRymw6umoZc6aWTEDyhqlzImEc8xWxcS4fF8khuoAe1WCYMijYUyZLKH7n4k6tqD7ollWJlNOXVB2WApOFNvmVQnTI8m9FpKZxQpg3%2FIlKuFErwXHIQ1VP7wnmm9MrF5CvY6A4lmvG5mCotpACGRUSUllNtBSCc8ITAfEFFEIYlfWlaxwprUaVX%2BxsKyuXd7ldDv2A2V0Yt8MngNn8OexYsKvB1GZ7D4jlu7xljwIx8bxYxleQ%2Bw8XaFsDN5g6p4YiQZPm6i9qLVx0cT0FXUbuvqouaJ7ZGd2xwROu9xV3%2BBa0lNJto9U6EVqeFVjaTICjKJUJNtE4UWmvklrgOClyj4jkqOkLB1mpQdlrAH%2BjuIC9NjZQp1d0rqq5hpFPNNyZgGxPrNZXXQNaE0PEpVnFltcYng2MVDQ%2FNscGFYyWhmhwLTsQxr%2FtEQAamkcGrvuXcWMhjMMvho4BSlJcKWq4LW%2FsK7x5ZXgRnYMYI45HACShmRFCYNxHrbd%2FqhjMcNKh3wpPGRh%2BKBk88lROcUJb78o6wFyLpFO%2FNj6DND7vXvZkHJ0jQIghJ5wkRWILn9VlUia6ACtdnD%2BgOWoNy0IZyBe%2BDQ7l%2FgfJmKOsbZwPLG3bz4FjWg5vRXsXtWYZTHbkTvPwJEE3qVy2lBPZNvQ19d9C%2BMrm0dS8LjXab1DNZ7WI%2BX%2FPW2gwnjyHebT7L4hQFtrrDHXrZu6zBvb3uOpI3GDj3C2jQjFfVUWzGq%2BBY8ar9An6JV3W80ukwM165p4pX9nbxiqb7xyuaXuLVvxWvDhqdNtIBMxqlOd8A%2FcUN4whxrN%2BRrDpeHPtYN%2BlDxzG3I46dKpGlB%2F9LHMuv0W799rUT2XahM5At%2F9fRsgqfx2T3%2Byf3%2BqXK7ciSHe1SVUH6Qu4ucnck0Gz%2FVOTuSKFtJvdeXHsLufeKJv89ub2O5N%2FxyO1cyP0Kuf02uRE6Fbn9NrnfY378tGRay6R7HenHo2XSEeresRtjl27117viS16Le9t%2B3T6rU%2F2OHMmhnArV%2BichRZvxsxtn%2FAc%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%2292ed73d6-4fbb-8d80-8b2c-9356339b3040%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%204%22%3E7VnLcpswFP0aZtJNhocheBkcp1m0M53Jos0qI4MCagRihGzjfH2vQOIR7IxJayaPeiFLV1evq3MO5tpwFmn5laM8%2Bc4iTA3bjErDuTJse%2B6aUErDrja4c782xJxEtclqDbfkCSujGhevSYSLnqNgjAqS940hyzIcip4Ncc62fbcHRvur5ijGA8NtiOjQ%2BpNEIqmtvn3R2m8wiRO9suXN654VCh9jztaZWs%2BwnYfqU3enSM%2BlDlokKGLbjslZGs6CMybqWlouMJWh1WGrx10f6G32zXEmjhngqm2InT46jiASqsm4SFjMMkSXrTWojoflBCa0EpFSqFpQxSURv6T53FWtO92TRZfyWqAZUlQUJKyN14Tqwb%2BxEDuFBLQWDEzt8t8Yy5XfA8vENUoJldC6wXSDBQkRdNQnkds%2FGAtlKtiah8rLVuhCPMbKa9bcA8AbsxQLvgMXjikSZNOfHSmgxY1fG2yoqHjvj71aeoPoWk1qQETnV7IMTANO61%2FoOpRBVS6HF0Yp0EJezDYhAt%2FmqDrbFojZv56DkUOUxJm8GggU5k0oN5gLXL4czGGY1ADHUwhXCjBTzW1LJ0uTIOlQSfv9TWAvpgI17KvTJZt3aoK3C%2FjZEPDeRICf7Qd8VQa2grcsZ4Zzadgehf0FKw61WNZoIYbGwVWDouayCmFAlGLKYo5ScMwxJ7Bjie9%2B34%2B2400zqKHLFBTSD%2BLP%2B2DwhjzxJ%2BKJN%2BSJwn2Ro0wjP4Roi%2FuKFGblZWovmL%2FruG9wCtA%2B3vtshC9F6SpC43ZTVnw34Szn1anOyi8jRhv2YtxysM7h%2BZ%2Bhvo%2Fpt6QPtt%2FXB9sc6kPj88%2F1wRpE6v8zttEOf6gd84m0wz9KOygpxP2WiOS%2B4ttrFOSJjFKQzsP7FJRuxPAjEtufktjOZ3%2Fwz4fkteyJ2Ds%2Fir0pKl%2FPWxg8irfPleJ0HH7EuzHTy4NP8aukNNxgxChrzBncq4%2BgV89eVJoXl45eOSd7UZksg%2FUef4hYe9Jb1lT5Lb14X81QKrGarQr5pd%2F%2FLcMPjswCvCyIo6g6RgyAqscy9cNmIGZTJvGsA9miyw5C6lSpW%2BHHG%2BrAu0iMuidM60Cz%2FSuh6uv8XeMs%2FwA%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%22d362a8da-704f-48d7-3e24-7edbcb737f6a%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%205%22%3E7VhLj5swEP41SN1LBWZJ4JjXdg%2BtVGkPbU%2BVFxxw12BknFd%2FfcdgXgG2oc1Gu21ysOzxzHg8M9%2BnGMNexPsPAqfRJx4QZiAz2Bv20kDIc0wYleBQCBzPLQShoEEhsmrBA%2F1JtFDbhRsakKylKDlnkqZtoc%2BThPiyJcNC8F1bbc1Z%2B9QUh6QjePAx60q%2F0EBGhdRF01p%2BT2gYlSdbE6%2FYecT%2BUyj4JtHnGche579iO8alL33RLMIB3zVE9sqwF4JzWczi%2FYIwldoybYXd3cBuFbcgiTzF4FaHIQ%2Fl1UkAmdBLLmTEQ55gtqql8%2Fx6RDkwYRXJmMHUginZU%2FlVid87evWt3EmkODS21PKbdkCSYKZKBkuf4SyjfiG8o6x0%2FINIedBdgjeSg6gO7SPnqdZb80Te4Zgy1Xb3hG2JpD6GjeKW6mqDedKijG%2BEr7WQ7jwsQqK17KpG0PqExwQuAiqCMCzptu0d6yYMK726EDDRteiviz56i9lGOzUgWd5SjXPTgNu603IO4zwfVwaaMAh0%2FihgFsrq4s3yMgYgUmXcRVSShxTnt90BjNvFHMwlZjRMVLEgdURUZ2yJkGT%2FfHq7idMG9kTjQfPFrV7uavBZJWSiBvBKvb9J9eQKgWEI2F0IOBeCgN0PgXycI93warw17Fm381kmT4AD8G%2BqppAGzBhhPBQ4BsWUCAoRq%2F5u732uN141giq4XAJC7hVCwxByuhCaXghCThdCGhJZipMSFAkJ82MzUM21zFIL%2FDcV%2B4zXlKkePt3g3QhdhuPHAI8LaJ%2BzAcDCmY%2BwskboGo7OEY4V4rXZmBjH6BpoMc45STYxEViSMUUBtry5GTQ4Ancbuq%2BJBdERCyKzy4L2S7Fg%2BXi50mAfDU67NOhdiAanJ9FgjPffoZPjP2FBsB2DtgbjvhwPPJHDGPfqym%2BfmP8FDnOPOMztclglOzuHoSuHDXOY1%2BUwa6CYZycxb%2FA5ZBnu%2FMRHUS9qlJtF42PCLHdp5XO3%2FZEB5IXmsnGsp3hAm8BYKDs6isIJMsvQQAE11Nx8Pml81kAN894TYXQKKjGmNWk7wBTwv2y61DxZB79sROK0LtjjxCydPEs6%2F8OD8piGqgfm72hocg4asvq7fdbo8GVZ0byB3uh3rp6knuuVDsv6O3K%2B1%2FhWb69%2BAQ%3D%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%221bfecccf-8d96-fd65-ab71-ac7026433fb6%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%206%22%3E7VtZb9s4EP41ApqHFjp9PMZOsn3YBRabAts%2BFbRFy9xSokHRidNfv0OJ1GHRsZRYitPaD4I4pHjM8c1wSFvePN79wdFm%2FRcLMbVcO9xZ3o3lutPAhqckPOWEYDrJCREnYU5ySsI9%2BYkVUX0XbUmI01pDwRgVZFMnLlmS4KWo0RDn7LHebMVofdQNinCDcL9EtEn9l4RinVMn7rikf8YkWuuRndE0r1mg5Y%2BIs22ixrNcb5X98uoY6b7UQtM1CtljheTdWt6cMybyt3g3x1SyVrMt%2F%2B7uQG0xb44T0eYDV01DPOml4xA4oYqMizWLWILobUmdZcvDsgMbSmsRU3h14BXviPgqyZ8CVfqmaxLBnypVsvhNdYCT8FqKDIpLitKULHPiHaG64%2F%2BwEE9KS9BWMCCVU%2FuTsY1qt2KJuEMxoVLtPmP6gAVZIqjIVymXVuNTyrZ8qUieUjPEI6xY5ze56RQyAtXHLMawEGjCMUWCPNR7R0oJo6JdKQh4UbIwy0XN5gHRrerUAmZNb%2BRzZluw2slYv8Nzlj1vLXdEYaKzBYe3SBQLr4qXUjAiKcbHNRH4foMyBjyCGdeFeZCXiJIokcICpmBejPGAucC7g2p4gHHqg0Dbg8ILXxUfS%2BNzdJN1xfB0u9ewulCG3m3gzfS8IYi66vtN1XcmAym6b1b07DlzlVrLp2951039pqloofSAshv5CnxAlGLKIo5iaLjBnMCMpRbX6%2F4uK87aTgqjGMJQnMEM5SydxREjCgxGZJuFeXIjCt7cW5xAtydv5wJGAyu2U1NrpeWFYicswTWttjtodYjSdTYr57AmV9V23FTb4O3CnnFTkZ%2FBckEQ%2FQfCf5REtIWOLpgQLFYFrjRIvouMe%2FZ%2BkxVwf84o46VMQOiCJNE8g3QpyY%2B59DgTwAgm4X5qn8Yc3FHdHjRvKvagJVU1h8nrhTBpCGHDcQjYKPCH3VVDILBEUWc0xyn5iRZZA8mNDSOJyCYUzKzgRvpG0OA0V2an4iopXsmuJN9gOHqtyJl8ZikIFnj%2FJRPWR%2F9EXB67n4I6n50GnycG2HFP4VD9PnDHqaBOxYN%2BLTzoIdw5M4eqcwE1j%2BoNBETO0C7hXcU6RtG4Q4nGNjiJPJIBiEh0LBOT5Du4g1hijqqGnqstTF%2FZHdpKKJMwMEKxxCf1GSDL0CPGaDfEUhvRoowxJ9lTvxsiytZjDC6waHCB6REHFVib1u3ltgeKfM3ixTY9HnwNuk1usUsu9g1Vp%2B6cxKu7DXziWGw5MNH%2BwoH0PlJye9uxojxIpqGZ%2FixYeIdo%2Bl546Dl7ijgakomGrIAGomSRbjSC51mCIMsSNLdbbxfdH4lZXrm5qgf9TuA2BGMCiJPgg2GTK%2BXgqwzN7yEBZ3976zf3t15PEjBkFS7BfRncTwzB%2FXSo4L6ZfDAFLDjZxpgjAfIAhfBUkNU23KGkPDVo0fxDh7bFxLr0T1NxddU64qrp4Tl5O38%2FZLANYZfdk7fTJ6YXmzba9LRp05X7Bj3b9LSVTRebohdYNHzbxeCq8NFhkzTvNqkf%2BKlL916XHRtF8SJE3eazy05Qu253nW67zgLHXrCX7IJsp8Ar%2F%2FgWx5j7PQleBRe8OoxXOsqr4ZU%2FEF7pwdsmGF%2BCVyS54NV54dU7jrv2cMwzZBn6w7HxBceewTHD1ar8tsEQOGa4W2UwghWhoIov20nl33aBsuIouU8kK%2FHyVzDvvUyJb7gZY0xWncS8pxfzfsa8RwbzHurmpB78iHlzlET4e7ot4xSwwO0Sd8lqvMR7oz4NfJGHBvmU5Inaotcgoc%2BllPDb3xhd2v4KiOnVEbO4NVvLLfeEmN7lluxziGlILntDJaIMN9v0ZXPHmsxaXjmvAOrve%2FF8ZMjt9naS6Rk25FJu1xVZVU8xDXe3zvJ4uMHUHs%2FYoVj%2Bqyyrq%2Fxzz7v9Hw%3D%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%22c48cefe7-a414-75a3-7076-9ca640ec0d0f%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%207%22%3E7Zhbb5swFIB%2FDVL3UgUIhDw2abo%2BbNKkPmx9qlw4AW8GI%2BPc9ut3DOZWhy1TW9QtzYNlHx9fzuWLsS13me4%2FCpInn3kEzHIm0d5yry3HmXsTLJXgUAm8eVAJYkGjSmS3gjv6E7RQj4s3NIKipyg5Z5LmfWHIswxC2ZMRIfiur7bmrL9qTmIwBHchYab0K41kUkkDZ9bKb4HGSb2y7c%2BrnkcS%2FogF32R6Pctx1%2BWv6k5JPZc2tEhIxHcdkbuy3KXgXFa1dL8Eplxbu60adzPQ2%2BxbQCZPGeDobchDbTpE6And5EImPOYZYatWuijNAzXBBFuJTBlWbazCnspvSnzp6dZ93ZNJceh0qea9ngCy6EqFDJshI0VBw0p4Q1k98XeQ8qCzhGwkR1G7tU%2Bc51pvzTN5Q1LKVNrdAtuCpCHBjspKZVrPTwXfiFCLXJ1mRMSgXTc1vWk3McLUB54CGoIqAhiRdNufnegkjBu9NhBY0bE4Hhe9my1hGz2phc6aX6tyMbHQ2mBW17FclOXKcnyGG108CqzFsjG8G17GECIVxl1CJdzlpHTADjHuB3PQl4TROFPBQqeAaNbYgpCwH0zDAcfpAa6vedD%2FF1Pd3LXw2TUySQe8Wu85rp6fNQJGjPpUTE0qvJEQmB5HoCwXjk54VU4t98rMfFbIE3DA%2F99cVdENhDFgPBYkRcUcBMUdq%2Fzu931pO940QQ0uYyBUH9uvz9Bb5cQzOfFH4sQzOdF5X%2BQkqzN%2FTUUhH1KaPWDOpqhfqk5qVVykq31sBhz7F9oXyN%2BHQf0nydJPhTeFldPHypmYWDWyF8fqvL%2FO%2FoCcbyI3Gwk5%2FyTkkIBLASnfwkWfvv%2BRi2BMLtxzP25mZu4HI%2BX%2B7KTcLwCv5tH7efNcrprPulG48t7Pm2HmApM5ezoSdIEJnb4L2VawOPFGdMpn4VIZ%2FZTd30J4Dper5nSrnyeOvE80Oi9O5cBF%2BKoT8OpdyCvTwTch%2FifefLxXvLFis31TLfs679bu6hc%3D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%22d2fcc3fd-1e0b-9973-6865-6f856db9f6cc%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%208%22%3E7VnJbtswEP0aAenFsFZLxzhxmkMLBMihzZGRGIktJQoUvfXrO5SolVbqIolqN%2FGBIGeG28x7Q1I27Kt095mjPPnKIkwNax7tDPvasKzAnUMpBftK4AZ%2BJYg5iSqR2QruyS%2BshKpfvCYRLnqGgjEqSN4XhizLcCh6MsQ52%2FbNnhjtz5qjGGuC%2BxBRXfqNRCKppL61aOW3mMRJPbPpBZXmEYU%2FY87WmZrPsOyn8lepU1SPpTZaJChi247IXhn2FWdMVLV0d4WpdG3ttqrfzYi2WTfHmTiqQ70Osa%2F3jiNwhWoyLhIWswzRVStdlvvDcoQ5tBKRUqiaUMU7Ir5L8cxVrYdakwm%2B76hk80ENgLPoUsYMmiFFRUHCSnhDaD3wDyzEXsEErQUDUbu0L4zlyu6JZeIGpYRK3N1iusGChAgU1S7l1kYdpUQFW%2FNQWdkKeojHWFn5TZAA%2B5ilGDYCJhxTJMimPzpSKIwbuzYSUFHBOBwYNfUG0bUa1ABnBdeyXM4N2K2%2FqOtQLstypQeTUuCMDNo2IQLf56jc2xZY2w%2FdqOcQJXEmQwOOwrxx5QZzgXfPO1N3k%2Bpgewp2Kj04qrltuWbWDEk6PKvtXuJY054K8aeKakdHtTsRqpvlvK7zO67v5JdG1SQbpTujwAQvDEzZFbaL9h2DnJFMFJ2R76SgQ0%2BzT0%2FtEPmDvevMB6CoVtBCpNnKUahxtFxIpDaDwpzNHA1ScKjmsprgHYKQQTxyzAlMKPNXLb2rRdaJJ0drPmV21L35zrKjq5PQmyg7uofP%2FLJcWuqEl6Vj2JeG5VFY3%2FKRQy0WZRi2unCUHOAGRCmmLOYoHVCkp7trFafNE39KnnhT8eQc782ezqHFRBzydA4pThQ5yjpUmaE8B2dd1GoYuGtxqFexTv%2FC%2BgIm%2BfRptMMAPX1snBTPnD7PGt51eea9Ec%2BsN3mfmsbohdE0zujCuNB55vyLC6M7gIjjPX9hHNoPLpgvvjAujkoCKRKc7Oo8INn6H5B1%2BLKe9GltfRyK42T1X5%2BsxwbG1%2FhQYR9kZXNuuEvDvT5jnDeYngTnwQfOx3Ee6Dg3R4L56kAP9MSvXlCm4S%2BPfEfV3Hi%2FTyl3UjaZh4N22QlU9dnbLcN44OV1Fh%2B5vTf8jgPN9j%2Bj6sLU%2Fi9nr34D%3C%2Fdiagram%3E%3Cdiagram%20id%3D%22c3c7a6a5-533a-bfe6-781f-eb7da2e91a92%22%20name%3D%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%209%22%3E7VpLc6M4EP41VM1cUkYYBh%2FjPDaH2aqpymFnTlMKKKBdgSghx3h%2F%2FbaQeOPETAIVT9YHSupuCam7v26pseVcJcUfAmfxnzwkzEKrsLCcawuhjbuCpyIcNMHd%2BJoQCRpqkt0Q7um%2FxBDNuGhHQ5J3BCXnTNKsSwx4mpJAdmhYCL7vij1y1n1rhiMyINwHmA2pf9FQxprqoy8N%2FY7QKK7ebHsbzXnAwT%2BR4LvUvM9CzmP50%2BwEV3OZjeYxDvm%2BRXJuLOdKcC51KymuCFOqrdSmx90e4dbrFiSVpwxYm2XIQ7V1EoImTJcLGfOIp5jdNNRtuT2iJlhBL5YJg6YNTVJQ%2BV2RL1zT%2B1FxUikOLZbq%2FjATkDS8VCaDbsBwntNAE28pqyb%2Bm0h5MF6Cd5IDqVnaV84zI%2FfIU3mLE8qU290R9kQkDTAw9C7V1o7qyZByvhOBkULG87CIiJFyahuB6xOeENgIiAjCsKRP3dmxccKolmsMAQ1ji3G7mFc%2FYbYzk1qgrM21em5XFuzW%2F1K14bktnzdDYzIGkFFG28dUkvsMl3vbA2i7pjuqOcxolCrTgKKIqFX5RIQkxfPKHKrJDHA84%2F0mOqxNd99Aza4AErdgVsm9RrG2PVTSPB7%2FXr3aGXq1vRq315u7tTPu1uVzi4wTq%2Bfaci4t5DFY4PZBQCtSLQifghZD%2BsCkEFYz1QRVYMYI45HACQhmRFBYtHLkLu9bw3jXUKlxsQRWNh8dKu4QKv5SCQDNovyW6lupuGbVednwzsgw9mtTczkU9osPLYGM01TmrZm%2FKUKDT%2BR38emseuetF%2BTd9arnFXoFjY%2FUWznJbdxhgDXRMs9wWsVLCgd1FTQ0B%2BZsM0cHpNPkBU7BoKfLf5ogO2UdFrqatnBG0inL1gnp8%2BejY47kpZgUGCDSy0iGWuci9N6SEdpcuB33rd25nY58e550hBY7up3jZcUfhkS01LHOPynqBJztkjKQKJEpqEzA7ecJJgwnDyGetpqiPJsCKNwykLrXcwYkE2FODTBdd39XwcPuZUpvGDrQbLe%2BeU5Tv0noqCzTjh3uQqGjevcLsSOh6c9SJG%2Fih7t9DnvnA4zeFQ%2BtliyH%2BP8D4xlgjNVK1kshY1gsSXChUVCBwMOJct30Ic%2BO5o9iSupscPY7pBzUu2yNFE9qtL05sharnpwlstYjyFos56yHOceUIW3L355cjKzB%2BHHrkd6iyWqkuqHsdtmylf5C4paW9IYIPIvvId7YrXoupVY1%2BbeNU7Z1tNBoW2dUaEQzHI1%2FqdDo9r%2BZec8XGvvyvcLkqwuNaOq5%2FQJnGZh0yuU8oZOKcbq%2BcHox7nwOLuuRL6RvdXCBbvN3A%2B0IzV86nJv%2FAA%3D%3D%3C%2Fdiagram%3E%3C%2Fmxfile%3E

### CODE REVIEW

File: 1.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# В диапазоне натуральных чисел от 2 до 99 определить, сколько из них кратны
```
Comment (Dzhoker1) [inline 1.py:1]: Вы сидите в чате Телеграм для группы? Там как раз обсуждали эту задачу. Как её решать. И я просил ваших однокурсников посмотреть урок с 2:04:00. Там я говорю о 8 ответах. У вас 1. 
Ход мыслей верный, а ответ не тот.

File: 2.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Во втором массиве сохранить индексы четных элементов первого массива.
+# Например, если дан массив со значениями 8, 3, 15, 6, 4, 2, то во второй
+# массив надо заполнить значениями 0, 3, 4, 5 (индексация начинается с нуля),
+# т.к. именно в этих позициях первого массива стоят четные числа.
+
+lst = input("Введите списко через запятую: ")
```
Comment (Dzhoker1) [inline 2.py:6]: Будем считать, что пользователь всё ещё идеальный )))

File: 3.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# В массиве случайных целых чисел поменять местами минимальный и максимальный элементы.
```
Comment (Dzhoker1) [inline 3.py:1]: lambda решение. Интересно. Спасибо.

File: 4.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Определить, какое число в массиве встречается чаще всего.
```
Comment (Dzhoker1) [inline 4.py:1]: Короткий python style. 
Специально делаете решения максимально короткими? Это не придирка. Просто интересно.

File: 5.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# В массиве найти максимальный отрицательный элемент.
```
Comment (Dzhoker1) [inline 5.py:1]: Решение понравилось

File: 6.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+min_item = min(enumerated, key=lambda x: x[1])
+
+
+def predicate(x):
+    return min_item[0] < x[0] < max_item[0] or \
+           min_item[0] > x[0] > max_item[0]
+
+
+filtered = filter(predicate, enumerated)
+range_sum = reduce(lambda a, b: a + b[1], filtered, 0)
```
Comment (Dzhoker1) [inline 6.py:22]: Даже reduce? ))) Сто лет его не видел. С тех пор как разработчики убрали его подальше в functools

File: 7.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# В одномерном массиве целых чисел определить два наименьших элемента.
+# Они могут быть как равны между собой (оба являться минимальными),
+# так и различаться.
+
+lst = input("Введите список через запятую: ")
+lst = list(map(int, lst.split(",")))
+
+first_min_item = min(lst)
+lst.remove(first_min_item)
```
Comment (Dzhoker1) [inline 7.py:9]: Решение есть, а одного элемента в списке нет. Но это мелочи )))

File: 8.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# Матрица 5x4 заполняется вводом с клавиатуры, кроме последних элементов строк.
```
Comment (Dzhoker1) [inline 8.py:1]: Отлично

File: 9.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Найти максимальный элемент среди минимальных элементов столбцов матрицы.
```
Comment (Dzhoker1) [inline 9.py:1]: Отлично

**Review by Dzhoker1 — COMMENTED** [review]
Оставил комментарии по коду

File: 1.py (hunk 1 above)
Comment (Buzyuka) [reply 1.py:1]: В чате я сижу, я там Яра, но у меня до 26 июля связь просто ужасная. Я сейчас на Алтае, в отпуске. Видео смотреть нет возможности, поэтому много интересного пропускаю.

File: 2.py (hunk 2 above)
Comment (Buzyuka) [reply 2.py:6]: Как раз издержки отсутствия нормальной связи. Много пропускаю.

File: 4.py (hunk 4 above)
Comment (Buzyuka) [reply 4.py:1]: Не то, чтобы специально, просто мне так удобно. Вообще, если честно, чем короче, тем проще рисовать блок-схему)) я с содроганием начинаю думать уже заранее, что опять надо блок-схему рисовать, поэтому стараюсь, чтобы решение было максимально коротким. Вы меня сделали счастливой, написав, что рисовать их больше не нужно. Из-за отсутствия связи этот момент я тоже пропустила))

File: 6.py (hunk 6 above)
Comment (Buzyuka) [reply 6.py:22]: Но ведь функционально же ;)

File: 1.py (hunk 1 above)
Comment (Buzyuka) [reply 1.py:1]: Пробовала сейчас посмотреть с телефона вебинар, но у меня на общий экран изображения нет. Вижу черный экран и слышу ваш голос. Так что с 4 уроком тоже посмотреть ничего не получится. А вот следующий урок я уже вернусь из отпуска и надеюсь смогу полноценно присутствовать на уроке.

**Pull request closed without merging**